source/code/plugin/filter_health_model_builder.rb

```diff
@@ -0,0 +1,223 @@
+                                )
+                        end
+
+                        instance_state = @state.get_state(monitor_instance_id)
+                        #puts "#{monitor_instance_id} #{instance_state.new_state} #{instance_state.old_state} #{instance_state.should_send}"
+                        should_send = instance_state.should_send
+
+                        # always send cluster monitor as a heartbeat
+                        if !should_send && monitor_instance_id != MonitorId::CLUSTER
+                            all_monitors.delete(monitor_instance_id)
+                        end
+                    }
+
+                    @log.info "after optimizing health signals all_monitors.size #{all_monitors.size}"
+
+                    # for each key in monitor.keys,
+                    # get the state from health_monitor_state
+                    # generate the record to send
+                    all_monitors.keys.each{|key|
+                        record = @provider.get_record(all_monitors[key], state)
+                        if record[HealthMonitorRecordFields::MONITOR_ID] == MonitorId::CLUSTER && all_monitors.size > 1
+                            old_state = record[HealthMonitorRecordFields::OLD_STATE]
+                            new_state = record[HealthMonitorRecordFields::NEW_STATE]
+                            if old_state != new_state && @cluster_old_state != old_state && @cluster_new_state != new_state
+                                    ApplicationInsightsUtility.sendCustomEvent("HealthModel_ClusterStateChanged",{"old_state" => old_state , "new_state" => new_state, "monitor_count" => all_monitors.size})
+                                    @log.info "sent telemetry for cluster state change from #{record['OldState']} to #{record['NewState']}"
+                                    @cluster_old_state = old_state
+                                    @cluster_new_state = new_state
+                            end
+                        end
+                        #@log.info "#{record["Details"]} #{record["MonitorInstanceId"]} #{record["OldState"]} #{record["NewState"]}"
+                        new_es.add(time, record)
+                    }
+
+                    @serializer.serialize(@state)
+                    @monitor_set = HealthModel::MonitorSet.new
+                    @model_builder = HealthModel::HealthModelBuilder.new(@hierarchy_builder, @state_finalizers, @monitor_set)
+
+                    router.emit_stream(@@rewrite_tag, new_es)
+                    # return an empty event stream, else the match will throw a NoMethodError
+                    return []
+                elsif tag.start_with?("oms.api.KubeHealth.AgentCollectionTime")
+                    # this filter also acts as a pass through as we are rewriting the tag and emitting to the fluent stream
+                    es
+                else
+                    raise 'Invalid tag #{tag} received'
```

Comment:
you are explicitly raising here. Who is handling ?

Reply:
the raise in line 217 below